crypto/mem.c

```diff
@@ -361,71 +198,145 @@ void *CRYPTO_realloc_clean(void *str, int old_len, int num, const char *file,
     if (str == NULL)
         return CRYPTO_malloc(num, file, line);
 
-    if (num <= 0)
+    if (num == 0) {
+        CRYPTO_clear_free(str, old_len);
         return NULL;
+    }
 
-    /*
-     * We don't support shrinking the buffer. Note the memcpy that copies
-     * |old_len| bytes to the new buffer, below.
-     */
-    if (num < old_len)
-        return NULL;
+    /* Can't shrink the buffer since memcpy below copies |old_len| bytes. */
+    if (num < old_len) {
+        memset((char*)str + num, 0, old_len - num);
+        return str;
+    }
+
+#ifdef CRYPTO_MDEBUG
+    if (call_malloc_debug) {
+        CRYPTO_mem_debug_realloc(str, NULL, num, file, line, 0);
+        ret = malloc(num);
```

Comment:
It might be worth a comment that malloc is used instead of realloc so that we can retain a handle on the old storage so as to overwrite it; it looks like a typo in a quick read, otherwise.